### PR TITLE
feat: initial role based authorization implementation + large refactor

### PR DESCRIPTION
**Pull request by darahayes**
## Motivation

(I'll fill this out on monday. Going home now 😄 )

## Checklist:

- [ ] Code has been tested locally by PR requester
- [ ] Changes have been successfully verified by another team member 

## Progress

- [x] Finished task
- [ ] TODO

## Additional Notes

PS.: Add images and/or .gifs to illustrate what was changed if this pull request modify the appearance/output of something presented to the users.

### CODE REVIEW

File: server/apolloServer.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      const context = {
+        request: req
+      }
+      if (AuthContextProvider) {
+        context.auth = new AuthContextProvider(req)
+      }
+      return context
+    },
+    tracing,
+    playground: {
```
Comment (darahayes) [inline server/apolloServer.js]: Now that I think of it, all this stuff could probably be refactored into `server/config/index.js`

Comment (wtrocki) [reply server/apolloServer.js]: LOL. This is what I have suggested, but forgot to click send on friday :)

**Comment by wtrocki** [comment]
Actually it will be worth now to update readme - that we allow to write our own providers etc.

**Comment by StephenCoady** [comment]
Great PR, really makes things cleaner. Have verified it is working but I notice you are still pushing. Ping me once you are done and I will re-verify.

**Comment by darahayes** [comment]
@StephenCoady it's ready to go now!

**Comment by darahayes** [comment]
So there are three things to come out of this PR.

1. Document in the Readme how to run the server locally, protected with keycloak, and how the login and token flow works to get it running with the playground. - https://issues.jboss.org/browse/AEROGEAR-7839
2. Investigate (how to) and implement integration tests for the whole auth piece. - https://issues.jboss.org/browse/AEROGEAR-7840
3. Create unit tests specifically for hasRole() - https://issues.jboss.org/browse/AEROGEAR-7841

**Pull request merged**